stream_alert/rule_processor/alert_forward.py

```diff
@@ -0,0 +1,199 @@
+                                    'S': datetime.now().strftime('%Y-%m-%dT%H:%M:%S.%fZ')
+                                },
+                                'Cluster': {'S': os.environ['CLUSTER']},
+                                'RuleDescription': {'S': alert['rule_description']},
+                                'Outputs': {'SS': alert['outputs']},
+                                # Compact JSON encoding (no extra spaces)
+                                'Record': {'S': json.dumps(alert['record'], separators=(',', ':'))},
+                                # TODO: Remove TTL after alert merger is implemented
+                                'TTL': {'N': str(int(time.time()) + 7200)}  # 2 hour TTL
+                            }
+                        }
+                    }
+                    for alert in batch
+                ]
+            }
+
+    def _batch_write(self):
+        """Write a batch of alerts to Dynamo, retrying with exponential backoff for failed items.
+
+        Returns:
+            (bool) True if *all* items were written successfully, False otherwise.
+        """
+        @backoff.on_predicate(backoff.expo,
+                              max_tries=self.BACKOFF_MAX_RETRIES, jitter=backoff.full_jitter,
+                              on_backoff=backoff_handlers.backoff_handler,
+                              on_success=backoff_handlers.success_handler,
+                              on_giveup=backoff_handlers.giveup_handler)
+        @backoff.on_exception(backoff.expo, ClientError,
+                              max_tries=self.BACKOFF_MAX_RETRIES, jitter=backoff.full_jitter,
+                              on_backoff=backoff_handlers.backoff_handler,
+                              on_success=backoff_handlers.success_handler,
+                              on_giveup=backoff_handlers.giveup_handler)
+        def decorated_batch_write(cls):
+            """batch_write_item with the unprocessed_items from the AlertForwarder instance.
+
+            There are 2 different errors to handle here:
+                (1) If Dynamo is unresponsive, a boto ClientError will be raised.
+                (2) The batch_write_item operation can fail halfway through, in which case the
+                    unprocessed items are returned in the response. In this case, unprocessed items
+                    are stored in the class instance, and we return False.
+                    The backoff.on_predicate will automatically retry with any Falsey value, and
+                    batch_write will run again, but only with the remaining unprocessed items.
+
+            Args:
+                cls (AlertForwarder): Instance of the AlertForwarder
+
+            Returns:
+                (bool) True if the batch write succeeded, False if there were UnprocessedItems.
+            """
+            response = cls.client_dynamo.batch_write_item(RequestItems=cls.unprocessed_items)
```

Comment:
IMHO we should shy away from using `cls` as the var name here and it should be reserved for use in classmethods. I was fairly confused when looking at this since I _thought_ this was within a classmethod... but I knew `unprocessed_items` was a instance property, so I couldn't wrap my head around it. Then I realized you were passing `self` as an arg and it made more sense :)

Reply:
Ah, good point. This is why code reviews are great - we can clarify anything which is confusing